### PR TITLE
Add KDF in counter mode ACVP Testing

### PR DESCRIPTION
**Pull request by skmcgrail**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

### CODE REVIEW

**Review by WillChilds-Klein — COMMENTED** [review]
code changes look good. how were they tested?

File: util/fipstools/acvp/acvptool/subprocess/kdf.go [hunk 1]
```diff
@@ -67,59 +106,114 @@ func (k *kdfPrimitive) Process(vectorSet []byte, m Transactable) (interface{}, e
...
-		// We send it as part of the response so that NIST can verify whether it is correct
-		fixedData := make([]byte, 4)
-		rand.Read(fixedData)
+	if group.OutputBits%8 != 0 {
+		return kdfTestGroupResponse{}, fmt.Errorf("%d bit key in test group %d: fractional bytes not supported", group.OutputBits, group.ID)
+	}
+
+	if group.CounterLocation != BeforeFixedDataCounterLocation {
+		// We only support the counter location being after fixed data
+		return kdfTestGroupResponse{}, fmt.Errorf("label location %q not supported", group.CounterLocation)
```
Comment (nebeid) [inline util/fipstools/acvp/acvptool/subprocess/kdf.go:185]: Same comment/question
```suggestion
		return kdfTestGroupResponse{}, fmt.Errorf("counter location %q not supported", group.CounterLocation)
```

File: util/fipstools/acvp/modulewrapper/modulewrapper.cc [hunk 2]
```diff
@@ -2865,6 +2903,13 @@ static struct {
     {"KDF/Feedback/HMAC-SHA2-512", 3, HKDF_expand<EVP_sha512>},
     {"KDF/Feedback/HMAC-SHA2-512/224", 3, HKDF_expand<EVP_sha512_224>},
     {"KDF/Feedback/HMAC-SHA2-512/256", 3, HKDF_expand<EVP_sha512_256>},
+    {"KDF/Counter/HMAC-SHA-1", 3, KBKDF_CTR_HMAC<EVP_sha1>},
+    {"KDF/Counter/HMAC-SHA2-224", 3, KBKDF_CTR_HMAC<EVP_sha224>},
+    {"KDF/Counter/HMAC-SHA2-256", 3, KBKDF_CTR_HMAC<EVP_sha256>},
+    {"KDF/Counter/HMAC-SHA2-384", 3, KBKDF_CTR_HMAC<EVP_sha384>},
+    {"KDF/Counter/HMAC-SHA2-512", 3, KBKDF_CTR_HMAC<EVP_sha512>},
+    {"KDF/Counter/HMAC-SHA2-512/224", 3, KBKDF_CTR_HMAC<EVP_sha512_224>},
+    {"KDF/Counter/HMAC-SHA2-512/256", 3, KBKDF_CTR_HMAC<EVP_sha512_256>},
```
[the remark this answers is not in the document]
Comment (skmcgrail) [reply util/fipstools/acvp/modulewrapper/modulewrapper.cc:2912]: Counter KDF specification technically support HMAC-SHA3, but AWS-LC's HMAC implementation does not support its usage currently. There is a separate internal discussions with some CryptoBR folk about whether this is something that should be added, or whether something like KMAC should be used instead.

**Comment by skmcgrail** [comment]
> code changes look good. how were they tested?

There is a CMake target that gets built and executed for FIPS that runs a Go program that executes the compressed test vector files included in the PR.

File: util/fipstools/acvp/acvptool/subprocess/kdf.go [hunk 3]
```diff
@@ -67,59 +106,114 @@ func (k *kdfPrimitive) Process(vectorSet []byte, m Transactable) (interface{}, e
...
+	if group.OutputBits%8 != 0 {
+		return kdfTestGroupResponse{}, fmt.Errorf("%d bit key in test group %d: fractional bytes not supported", group.OutputBits, group.ID)
+	}
 
-		if group.CounterBits != 8 {
-			// We only support counter lengths of 8
-			return nil, fmt.Errorf("counter length %q not supported", group.CounterBits)
+	if group.CounterLocation != AfterFixedDataCounterLocation {
+		// We only support the counter location being after fixed data
+		return kdfTestGroupResponse{}, fmt.Errorf("label location %q not supported", group.CounterLocation)
```
[the remark this answers is not in the document]
Comment (skmcgrail) [reply util/fipstools/acvp/acvptool/subprocess/kdf.go:140]: Ah good catch, I think I reused an error string and forgot to fix this :)

**Review by WillChilds-Klein — APPROVED** [review]

**Review by nebeid — APPROVED** [review]

**Comment by skmcgrail** [comment]
Fuzzing passed, but one job status is hung in the incorrect state.

**Pull request merged**